### DISCUSSION TITLE
OpenModelica and OMEdit in WSL

### DISCUSSION THREAD

**Discussion by thorade** [OpenModelica/OpenModelica#11408] [General]
Windows Subsystem for Linux is a convenient way to run Linux applications from Windows.
https://learn.microsoft.com/en-us/windows/wsl/about

It is also supported to run apps with GUI from Windows:
https://learn.microsoft.com/en-us/windows/wsl/tutorials/gui-apps
https://ubuntu.com/tutorials/install-ubuntu-on-wsl2-on-windows-11-with-gui-support

Is anybody using that to run OpenModelica and OMEdit? What are the experiences so far?

**Comment by AnHeuermann** [comment]
I'm using it to build and develop OpenModelica and a lot of other projects. I'm working mostly on the shell or in VS Code, but the GUI apps usually work as well. Only from time to time minor issues creep up, like https://github.com/OpenModelica/OpenModelica/issues/11029 or random crashes.
Compiling code is also faster than using the provided MINGW environment on Windows (everything is checked by Microsoft Defender which is good and bad).

If you are developing OpenModelica itself or other applications around OpenModelica I would highly recommend WSL 2 (if you are somewhat familiar with Linux). If you are only using OMEdit to model/simulate Modelica models I would probably stick to the Windows version, it looks better on high resolution screens that the WSL GUI apps.

**Comment by casella** [comment]
@thorade if you play around with it and you are also happy, we could eventually make this somewhat "official".

**Comment by mahge** [comment]
I have also been using WSL for more than 2 years now. It was not the most straightforward thing for GUI apps in the early days. Lately, using the updated WSL2, I can't remember the last time I had any significant issues with OMEdit or other GUI tools.

I have Ubuntu, fedora, and Arch Linux and I compile and use OpenModelica on all of them without any issues.

**Reply by sjoelund** [reply]
With Windows 11 at home, the GUI is even easier (without needing an X server installed). It just works...

**Reply by mahge** [reply]
That is right. I think switching to Windows 11 was one of the changes I did. I don't remember what exactly it was but some feature was not available in Windows 10 and had to switch to Windows 11. I think it was indeed related to GUIs. Not a bad thing since Windows 11 feels better than 10 so far even though I can not point out why :)

**Comment by casella** [comment]
Good to hear! 

We should probably start to suggest this as an option on the Windows download page. However, I guess we should somehow test this in a systematic way before we declare officially that we support WSL.

@mahge, @adrpo what do you think?

**Reply by mahge** [reply]
Technically speaking there is nothing to support specifically for WSL2. OpenModelica has support for Linux and Microsoft Windows has support for Linux. In this setup, as far as we care concerned, OpenModelica is running on a Linux machine.

Running OpenModelica on WSL2 should behave exactly as running OpenModelica on the same Linux distribution being used in WSL2. GUI apps might look deferent when run through WSL2 compared to native Windows apps but they should look similar to what they look like when they are run in the specific Linux distribution. 

If there are issues with this setup then either

1. The issue is in OpenModelica Linux compilation/packaging/usage itself which we have to fix for our Linux distributions (i.e., they are not specifically WSL2 issues) 
2. It is an issue with WSL2, in which case it is up to Microsoft to fix them. 

We can mention that OpenModelica runs fine in WSL2 but, in my opinion, we should not suggest it as an alternative to the Windows version. We are suggesting people run Linux on Windows just for OpenModelica 🙂. We are suggesting that people download and install Ubuntu just to use OpenModelica when we already have a native Windows version.

**Reply by sjoelund** [reply]
The only thing needed to work in WSL is that you don't depend on any specific devices in /dev. Like formatting partitions, trying to use webcams, etc. But I don't see a major need for this in OpenModelica so it should keep working for a long time.

Regardless, I think most of the Windows developers use WSL2 most of the time.

**Reply by casella** [reply]
Do they?

**Comment by casella** [comment]
> Technically speaking there is nothing to support specifically for WSL2. OpenModelica has support for Linux and Microsoft Windows has support for Linux. In this setup, as far as we care concerned, OpenModelica is running on a Linux machine.

True, but we should check that it works fine, even though it is not our responsibility to make sure it does.

> We can mention that OpenModelica runs fine in WSL2 but, in my opinion, we should not suggest it as an alternative to the Windows version. 

Why not? For example, the Linux version is so much faster when compiling small models, because it doesn't have the burden of the antivirus/antimalware checks. If one already uses WSL2 for some reason (non of our business), it's good to know that he/she can also use OMC in there.
